src/main/java/org/veupathdb/lib/container/jaxrs/server/middleware/AuthFilter.java

```diff
@@ -69,15 +70,73 @@ public AuthFilter(Options opts) {
     if (opts.getAuthSecretKey()
         .filter(secretKey -> !secretKey.isBlank())
         .isEmpty())
-      throw new InvalidConfigException("Auth secret key is required for this "
-        + "service");
+      throw new InvalidConfigException("Auth secret key is required for this service");
   }
 
-  public static Optional<String> findAuthValue(ContainerRequestContext req) {
+  @Override
+  public void filter(ContainerRequestContext req) {
+    log.trace("AuthFilter#filter");
+
+    // Determine what the auth requirements and/or allowances are for the target resource
+    var requirements = fetchAuthRequirements();
+
+    // look for admin and check valid (500/403), but only if needed
+    boolean hasValidAdmin = requirements.adminDiscoveryRequired && hasValidAdminAuth(req);
+
+    // return 403 response if admin required but not present
+    if (requirements.adminRequired && !hasValidAdmin)
+      req.abortWith(build403());
+
+    // return now if no need to look up user information
+    if (!requirements.userDiscoveryRequired)
+      return;
+
+    // look for user submitted using normal auth (will abort with 401 if present but invalid, 500 if error)
+    Optional<User> authUser = findAuthUser(req);
+
+    // check if admin override need not be considered
+    if (!hasValidAdmin || requirements.overrideOption == AdminOverrideOption.DISALLOW) {
+
+      // user required but not present
+      if (authUser.isEmpty())
+        req.abortWith(build401());
+
+      // non-guest user required but user is guest
+      if (!requirements.guestsAllowed && authUser.orElseThrow().isGuest())
+        req.abortWith(build403());
+
+      // set request_user user to submitted value
+      req.setProperty(Globals.REQUEST_USER, authUser);
+    }
+
+    // valid admin wishes to override user ID and apply a custom user ID
+    else { // i.e. hasValidAdmin && override != disallow
+
+      // find proxied user
+      Optional<User> proxiedUser = findProxiedUser(req);
+
+      // if override is only allowed with proxied user but user not present, return 401
+      if (requirements.overrideOption == AdminOverrideOption.ALLOW_WITH_USER && proxiedUser.isEmpty())
+        req.abortWith(build401());
+
+      // set proxied user as the "request user" (may be empty)
+      req.setProperty(Globals.REQUEST_USER, proxiedUser);
+    }
+  }
+
+  private AuthRequirements fetchAuthRequirements() {
+    var ref = resource.getResourceClass().getCanonicalName() + "#" + resource.getResourceMethod().getName();
+    if (!CACHE.containsKey(ref)) {
+      CACHE.put(ref, new AuthRequirements(resource));
+    }
+    return CACHE.get(ref);
```

Comment:
Could use `computeIfAbsent` here to consolidate to a single call. Doesn't really matter, but since it's a synchronized map it might be worth it.